04_Scheduling/03_Processes_And_Threads.md

```diff
@@ -0,0 +1,385 @@
+
+We'll also need to keep track of the thread's current status, and you may want some place to keep flags of your own (is it a kernel thread vs user thread etc).
+
+### Changes Required
+
+Let's look at what our `thread_t` structure will need:
+
+```c
+typdef struct {
+    size_t tid;
+    cpu_status_t* context;
+    status_t status;
+    char* name[THREAD_NAME_LEN];
+    thread_t* next;
+} thread_t;
+```
+
+The `status_t` struct is the same one previously used for the proceses, but since we are scheduling threads now, we'll use it for the thread.
+
+You might be wondering where the stack is stored, and it's actually the `context` field. You'll remember that we store the current context on the stack when an interrupt is served, so this field actually represents both the stack and the context.
+
+We'll also need to adjust our process, to make use of threads:
+
+```c
+typedef struct {
+    size_t pid;
+    thread_t* threads;
+    void* root_page_table;
+    char name[NAME_MAX_LEN];
+} process_t;
+```
+
+We're going to use a linked list as our data structure to manage threads. Adding a new thread would look something like the following:
+
+```c
+size_t next_thread_id = 0;
+
+thread_t* add_thread(process_t* proc, char* name, void(*function)(void*), void* arg)
+{
+    thread_t* thread = malloc(sizeof(thread_t));
+    if (proc->threads = NULL)
+        proc->threads = thread;
+    else
+    {
+        for (thread_t* scan = proc->threads; scan != NULL; scan = scan->next)
+        {
+            if (scan->next != NULL)
+                continue;
+            scan->next = thread;
+            break;
+        }
+    }
+
+    strncpy(thread->name, name, NAME_MAX_LEN);
+    thread->tid = next_thread_id++;
+    thread->status = READY;
+    thread->next = NULL:
+    thread->context.iret_ss = KERNEL_SS;
+    thread->context.iret_rsp = alloc_stack();
+    thread->context.iret_flags = 0x202;
+    thread->context.iret_cs = KERNEL_CS;
+    thread->context.iret_rip = (uint64_t)function;
+    thread->context.rdi = (uint64_t)arg;
+    thread->context.rbp = 0;
+
+    return thread;
+}
+```
+
+You'll notice this function looks almost identical to the `create_process` function from before. That's because a lot of it is the same! The first part of the function is just inserting the new thread at the end of the list of threads.
+
+Let's look at how our `create_process` function would look now:
+
+```c
+process_t* create_process(char* name)
+{
+    process_t* process = alloc_process();
+    process->pid = next_process_id++;
+    process->threads = NULL;
+    process->root_page_table = vmm_create();
+    strncpy(process->name, name, NAME_MAX_LEN);
+    return process;
+}
+```
+
+The `alloc_process` function is just a placeholder, for brevity, `vmm_create` is also a placeholder. Replace these with your own code.
```

Comment:
even if iti quite clear just spending a sentence to explain what are the two function expected output, especially for the vmm_create can be useful?